File: src/entities/vocabulary.rs
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use typed_builder::TypedBuilder;
use uuid::Uuid;

/// Represents a vocabulary word.
#[derive(sqlx::FromRow, Debug, Clone, Serialize)]
pub struct Vocabulary {
    pub id: Uuid,
    pub created_at: OffsetDateTime,
    pub updated_at: OffsetDateTime,

    /// The vocabulary name/meaning.
    pub name: String,
    /// The vocabulary level.
    pub level: i32,
    /// Alternative names for the vocabulary.
    pub alt_names: Vec<String>,
    /// The vocabulary writing.
    pub word: String,
    /// The vocabulary word type (e.g: noun, verb)
    pub word_type: Vec<String>,
    /// The word reading.
    pub reading: String,
    /// Mnemonics to help you remember the vocabulary meaning.
    /// They are stored in a string using markdown syntax.
    pub meaning_mnemonic: String,
    /// Mnemonics to help you remember the vocabulary reading.
    /// They are stored in a string using markdown syntax.
    pub reading_mnemonic: String,
    /// Vocabulary name synonyms defined by the user.
    pub user_synonyms: Option<Vec<String>>,
    /// User notes on this kanji meaning.
    pub user_meaning_note: Option<String>,
    /// User notes on this kanji reading.
    pub user_reading_note: Option<String>,
}

#[derive(Debug, Clone, TypedBuilder, Deserialize)]
pub struct InsertVocabulary {
    #[builder(setter(into))]
    pub name: String,
    pub level: i32,
    #[builder(default)]
    pub alt_names: Vec<String>,
    #[builder(setter(into))]
    pub word: String,
    pub word_type: Vec<String>,
    #[builder(setter(into))]
    pub reading: String,
    #[builder(setter(into))]
    pub meaning_mnemonic: String,
    #[builder(setter(into))]
    pub reading_mnemonic: String,
    pub kanji_composition: Vec<String>,
}

#[derive(Debug, Clone, TypedBuilder, Deserialize)]
pub struct GetVocabulary {
    pub word: String,
}

File: backend/core/domain/cards/vocabulary.go
package cards

import (
	"time"

	"sixels.io/manekani/services/cards/ent/schema"

	"github.com/google/uuid"
)

type (
	Vocabulary struct {
		Id              uuid.UUID         `json:"id"`
		CreatedAt       time.Time         `json:"created_at"`
		UpdatedAt       time.Time         `json:"updated_at"`
		Name            string            `json:"name"`
		Level           int32             `json:"level"`
		AltNames        []string          `json:"alt_names,omitempty"`
		Word            string            `json:"word"`
		WordType        []string          `json:"word_type"`
		Reading         string            `json:"reading"`
		AltReadings     []string          `json:"alt_readings,omitempty"`
		Patterns        []schema.Pattern  `json:"patterns"`
		Sentences       []schema.Sentence `json:"sentences"`
		MeaningMnemonic string            `json:"meaning_mnemonic"`
		ReadingMnemonic string            `json:"reading_mnemonic"`
	}

	CreateVocabularyRequest struct {
		Name             string            `json:"name" form:"name"`
		Level            int32             `json:"level" form:"level"`
		AltNames         []string          `json:"alt_names,omitempty" form:"alt_names"`
		Word             string            `json:"word" form:"word"`
		WordType         []string          `json:"word_type" form:"word_type"`
		Reading          string            `json:"reading" form:"reading"`
		AltReadings      []string          `json:"alt_readings,omitempty" form:"alt_readings"`
		MeaningMnemonic  string            `json:"meaning_mnemonic" form:"meaning_mnemonic"`
		ReadingMnemonic  string            `json:"reading_mnemonic" form:"reading_mnemonic"`
		Patterns         []schema.Pattern  `json:"patterns" form:"patterns"`
		Sentences        []schema.Sentence `json:"sentences" form:"sentences"`
		KanjiComposition []string          `json:"kanji_composition,omitempty" form:"kanji_composition"`
	}

	UpdateVocabularyRequest struct {
		Name             *string            `json:"name,omitempty" form:"name"`
		Level            *int32             `json:"level,omitempty" form:"level"`
		AltNames         *[]string          `json:"alt_names,omitempty" form:"alt_names"`
		WordType         *[]string          `json:"word_type,omitempty" form:"word_type"`
		Reading          *string            `json:"reading,omitempty" form:"reading"`
		AltReadings      *[]string          `json:"alt_readings,omitempty" form:"alt_readings"`
		MeaningMnemonic  *string            `json:"meaning_mnemonic,omitempty" form:"meaning_mnemonic"`
		ReadingMnemonic  *string            `json:"reading_mnemonic,omitempty" form:"reading_mnemonic"`
		Patterns         *[]schema.Pattern  `json:"patterns,omitempty" form:"patterns"`
		Sentences        *[]schema.Sentence `json:"sentences,omitempty" form:"sentences"`
		KanjiComposition *[]string          `json:"kanji_composition,omitempty" form:"kanji_composition"`
	}

	PartialVocabularyResponse struct {
		Id       uuid.UUID `json:"id"`
		Name     string    `json:"name"`
		AltNames []string  `json:"alt_names"`
		Reading  string    `json:"reading"`
		Word     string    `json:"word"`
		Level    int32     `json:"level"`
	}
)
